### PR TITLE
Integration of individual components

### LINKED ISSUE #13: Implement LQR controller framework
Implement LQR controller framework

### LINKED ISSUE #22: Possible bug in quaternion difference
## Description of the bug 

In MATLAB, I defined two quaternions and the computed their difference as follows:

```matlab
p = angle2quat(deg2rad(10), deg2rad(3), deg2rad(-5));
q = angle2quat(deg2rad(5), deg2rad(4), deg2rad(9));
d = quatmultiply(p, quatconj(q))
```

The result is $d = (0.9912,   -0.1195,   -0.0245,    0.0507)$.

I did the same using our C++ code and got the result $d' = (0.991239, 0.124392, -0.0244649, 0.0507255)$. The $w$, $y$ and $z$ values are the same. The $x$-coordinate is not.

## Note

There is a little chance that MATLAB has a different definition of the multiplication of quaternions - we need to double-check - so this is not certainly a bug, but we need to check.

## To reproduce

Run [this code](https://gist.github.com/alphaville/bcb97e44e33890cc7afaefccc181e5da). You can do this using [this online C++ compiler](https://www.programiz.com/cpp-programming/online-compiler/).

### PR DESCRIPTION
**Pull request by jamie-54**
## Main Changes

- new quaternion class implementation
- first version of controller
- getter for angular velocity in AHRS
- `main.cpp`: experimental integration

## Associated Issues

- Closes #13
- Closes #22

### CODE REVIEW

**Review by alphaville — COMMENTED** [review]
Jamie, I made a few edits directly to this PR. Let's discuss tomorrow morning.

**Comment by jamie-54** [comment]
> Jamie, I made a few edits directly to this PR. Let's discuss tomorrow morning.

Ok I'm in uni on 10th floor now

**Review by alphaville — COMMENTED** [review]
Overall it looks good. Just have a look at #25 first.

**Comment by jamie-54** [comment]
Hi all, flight-zero is now working correctly like last Friday and the correct code is now pushed to this branch.
@alphaville I Think we might be better to merge this working code into dev first and then we can sort issue https://github.com/QUB-ASL/bzzz/pull/25. Since I had to go back to a previous commit there is a few other parameters and things need to be defined again which we can sort with along with https://github.com/QUB-ASL/bzzz/pull/25. Thanks

**Comment by jamie-54** [comment]
Hi all, while this current version of the code seems to work very well as seen on Friday, I thought I'd increase `PITCH_MAX_RAD` in the `config.hpp` file to see issue https://github.com/QUB-ASL/bzzz/issues/24 is still happening, and unfortunately it is.
Any thoughts can we discuss under https://github.com/QUB-ASL/bzzz/issues/24

File: src/main.cpp [hunk 1]
```diff
@@ -1,43 +1,161 @@
...
+}
+
+/**
+ * Setup the AHRS
+ */
+void setupAHRS()
+{
+  ahrs.setup();
+  ahrs.preflightCalibrate(false);
+  // TODO create #define's with these parameters
```
Comment (alphaville) [inline src/main.cpp:87]: Use `#define` for theses parameters

File: src/main.cpp [hunk 2]
```diff
@@ -1,43 +1,161 @@
...
+      radio.rollReferenceAngleRad());
+
+  Quaternion currentQuaternion(quaternionImuData);
+  Quaternion relativeQuaternion = currentQuaternion - initialQuaternion;
+  Quaternion attitudeError = referenceQuaternion - relativeQuaternion; // e = set point - measured
+
+  controller.controlAction(attitudeError, angularVelocity, controls);
+
+  // TODO Remove hard-coded numbers from here
+  float throttleFromRadio = radio.throttleReferencePercentage() * 1000 + 1000;
```
Comment (alphaville) [inline src/main.cpp:140]: Use `#define` for theses parameters

File: src/radio.cpp [hunk 3]
```diff
@@ -1,122 +1,149 @@
...
-
         }
     }
 
     void Radio::readRadioData(
-        int &radioThrottle, 
-        int &radioRoll, 
-        int &radioPitch, 
-        int &radioYawRate, 
+        int &radioThrottle,
```
Comment (alphaville) [inline src/radio.cpp:51]: Shall we remove this method?

**Review by alphaville — CHANGES_REQUESTED** [review]
Looks good, but have a look at a couple of comments I made. Before we merge this:

- do you want to address #26 now?
- what do you think about #25?

**Comment by alphaville** [comment]
@jamie-54 since it flies, let's merge this into `main` and create a tag. Do you want to do some final changes first?

**Comment by jamie-54** [comment]
Hi I've just pushed the code that works and yea think we can merge it then tidy up the code and check that it still works thanks

**Review by alphaville — APPROVED** [review]
@jamie-54, very good - let's merge this now (first into `dev` and then `dev` into `main`) and we'll take it from there

**Comment by jamie-54** [comment]
ok

**Pull request merged**